mlir/lib/Quantum/Transforms/wrapper_for_wrapper.cpp

```diff
@@ -0,0 +1,220 @@
+        if (callee) {
+            auto stringRef = callee.value().data();
+            Operation *fnDecl = moduleOp.lookupSymbol<LLVM::LLVMFuncOp>(stringRef);
+            retval = (bool)(fnDecl->getAttrOfType<UnitAttr>(
+                LLVM::LLVMDialect::getEmitCWrapperAttrName()));
+        }
+    });
+    return retval;
+}
+
+struct WrapperForWrapperTransform : public OpRewritePattern<LLVM::LLVMFuncOp> {
+    using OpRewritePattern<LLVM::LLVMFuncOp>::OpRewritePattern;
+
+    LogicalResult match(LLVM::LLVMFuncOp op) const override;
+    void rewrite(LLVM::LLVMFuncOp op, PatternRewriter &rewriter) const override;
+};
+
+LogicalResult WrapperForWrapperTransform::match(LLVM::LLVMFuncOp op) const
+{
+    std::string _mlir_ciface = "_mlir_ciface_";
+    size_t _mlir_ciface_len = _mlir_ciface.length();
+    auto symName = op.getSymName();
+    const char *symNameStr = symName.data();
+    size_t symNameLength = strlen(symNameStr);
+    // Filter based on name.
+    if (symNameLength <= _mlir_ciface_len)
+        return failure();
+
+    bool nameMatches = 0 == strncmp(symNameStr, _mlir_ciface.c_str(), _mlir_ciface_len);
+    if (!nameMatches)
+        return failure();
+
+    // Function must only contain a single function call.
+    size_t callsites = countCallsites(op);
+    if (callsites != 1)
+        return failure();
+
+    // The only call must contain the emitCWrapper attribute
+    bool hasAttribute = callsiteHasAttribute(op);
+
+    return hasAttribute ? success() : failure();
+}
+
+static bool functionHasReturns(LLVM::LLVMFuncOp op)
+{
+    bool result = false;
+    op.walk([&](LLVM::CallOp callOp) { result = (bool)callOp.getResult(); });
+    return result;
+}
+
+static bool functionHasInputs(LLVM::LLVMFuncOp op)
+{
+    bool result = false;
+    op.walk([&](LLVM::CallOp callOp) { result = !callOp.getOperandTypes().empty(); });
+    return result;
+}
+
+static LLVM::LLVMFunctionType
+convertFunctionTypeCatalystWrapper(PatternRewriter &rewriter, LLVM::LLVMFunctionType functionType,
+                                   bool hasReturns, bool hasInputs)
+{
+    SmallVector<Type, 2> transformedInputs;
+
+    Type ptr = LLVM::LLVMPointerType::get(rewriter.getContext());
+    auto inputs = functionType.getParams();
+    Type resultType = hasReturns ? inputs.front() : ptr;
+    transformedInputs.push_back(resultType);
+
+    if (hasReturns) {
+        inputs = inputs.drop_front();
+    }
+
+    LLVMTypeConverter typeConverter(rewriter.getContext());
+    Type inputType = hasInputs ? typeConverter.packFunctionResults(inputs) : ptr;
+    bool noChange = inputs.size() == 1;
+    if (noChange) {
+        // Still wrap the pointer into a struct
+        // for uniformity in Python and in the unwrapping.
+        inputType = LLVM::LLVMStructType::getLiteral(rewriter.getContext(), inputType);
+    }
+    if (auto structType = inputType.dyn_cast<LLVM::LLVMStructType>()) {
+        inputType = LLVM::LLVMPointerType::get(structType);
+    }
```

Comment:
Is it ever not a struct?

Reply:
Ah got it. Why the conversion though, is `if (inputType.isa<LLVM::LLVMStructType>())` not sufficient?

Reply:
https://github.com/PennyLaneAI/catalyst/pull/106/commits/bd8b47576f80932d16a24ac5f6314d38b5a388a7